### PR TITLE
[SPARK-12642][SQL] improve the hash expression to be decoupled from unsafe row

### PR DESCRIPTION
**Pull request by cloud-fan**
https://issues.apache.org/jira/browse/SPARK-12642

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @nongli @rxin

File: unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java [hunk 1]
```diff
@@ -51,16 +55,38 @@ public int hashUnsafeWords(Object base, long offset, int lengthInBytes) {
   public static int hashUnsafeWords(Object base, long offset, int lengthInBytes, int seed) {
     // This is based on Guava's `Murmur32_Hasher.processRemaining(ByteBuffer)` method.
     assert (lengthInBytes % 8 == 0): "lengthInBytes must be a multiple of 8 (word-aligned)";
+    int h1 = hashBytesByInt(base, offset, lengthInBytes, seed);
+    return fmix(h1, lengthInBytes);
+  }
+
+  public static int hashUnsafeBytes(Object base, long offset, int lengthInBytes, int seed) {
```
Comment (nongli) [inline unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java]: what's the purpose of this one vs the other ones? Can we unify? I can't imagine someone caring if we hash byte by byte vs hashing the initial bytes in large chunks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -206,22 +230,230 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
...
+      case d: Double => hashLong(java.lang.Double.doubleToLongBits(d))
+      case d: Decimal =>
+        val precision = dataType.asInstanceOf[DecimalType].precision
+        if (precision <= Decimal.MAX_LONG_DIGITS) {
+          hashLong(d.toUnscaledLong)
+        } else {
+          val bytes = d.toJavaBigDecimal.unscaledValue().toByteArray
+          Murmur3_x86_32.hashUnsafeBytes(bytes, Platform.BYTE_ARRAY_OFFSET, bytes.length, seed)
+        }
+      case c: CalendarInterval => hashInt(c.months) ^ hashLong(c.microseconds)
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I don't think you should mix it this way. This should call into murmur twice:
hashInt(c.months, hashInt(c.microseconds, seed))

File: unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java (hunk 1 above)
Comment (cloud-fan) [inline unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java]: `hashUnsafeWords` requires the bytes word-aligned, and this one breaks this limitation. It's useful to hash binary type and string type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I haven't got the intuition to measure a hash algorithm, could you provide more information? And it'll be great if you can also take a look at array type and map type, thanks!

Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: My general thinking is to pick a hash function that is known to be good, e.g. murmur3 and make sure what I do  when I use it doesn't have problems: I don't convert NULLs to a common value (e.g. null map to 0 would be bad) or introduce logic if I don't have to. I *think* what you have is probably okay but something similar like hashInt(c.months) ^ hashInt(c.day) would compute 0 whenever months == day. I try not to reason about this by delegating to the carefully crafted hash functions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 3]
```diff
@@ -206,22 +230,230 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
...
+
+      case array: ArrayData =>
+        val elementType = dataType match {
+          case udt: UserDefinedType[_] => udt.sqlType.asInstanceOf[ArrayType].elementType
+          case ArrayType(et, _) => et
+        }
+        var result = 0
+        var i = 0
+        while (i < array.numElements()) {
+          val hashValue = computeHash(array.get(i, elementType), elementType, seed)
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I think this is better as

var result = seed
while () 
  result = computeHash(..., result)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 4]
```diff
@@ -206,22 +230,230 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
...
+            mapType.keyType -> mapType.valueType
+          case MapType(kt, vt, _) => kt -> vt
+        }
+        val keys = map.keyArray()
+        val values = map.valueArray()
+        var result = 0
+        var i = 0
+        while (i < map.numElements()) {
+          val keyHash = computeHash(keys.get(i, kt), kt, seed)
+          val valueHash = computeHash(values.get(i, vt), vt, seed)
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Same here. Think of the hash more as the intermediate value and murmur will mix it up for you.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 5]
```diff
@@ -206,22 +232,225 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
...
+        var result = seed
+        var i = 0
+        while (i < map.numElements()) {
+          result = computeHash(keys.get(i, kt), kt, result)
+          result = computeHash(values.get(i, vt), vt, result)
+          i += 1
+        }
+        result
+
+      case struct: InternalRow =>
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: why is this different than just calling eval()

**Comment by nongli** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: This maybe a nested struct and is different from `input` needed by this hash expression.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Actually it's possible to abstract it and reduce duplicated code, I'll do it in follow-up PR

**Comment by rxin** [comment]
I'm going to merge this. Thanks.

**Pull request closed without merging**